### PR TITLE
Enable torch.eye for bool and half

### LINKED ISSUE #24088: RuntimeError: "eye" not implemented for 'Bool' when on CPU
## 🐛 Bug
torch.eye doesn't work for dtype torch.bool when the device is CPU, but it does work on GPU. 

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "eye" not implemented for 'Bool'

>>> torch.eye(5, dtype=torch.bool, device=torch.device("cuda"))
tensor([[ True, False, False, False, False],
        [False,  True, False, False, False],
        [False, False,  True, False, False],
        [False, False, False,  True, False],
        [False, False, False, False,  True]], device='cuda:0')

## Expected behavior
I was expecting torch.eye for bool to be implemented for both CPU and CUDA. Maybe there are other torch functions with this issue as well?

## Environment
Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1060
Nvidia driver version: 430.40
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.17.0
[pip3] torch==1.2.0
[pip3] torchvision==0.4.0
[conda] Could not collect

 - PyTorch Version (e.g., 1.0): 1.2
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Python version: 3.6.8
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: GTX 1060 (6GB)

## Additional context
I can workaround on CPU for now by casting to torch.bool after calling torch.eye.

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- Enable torch.eye for bool and float16 dtypes

Test Plan:
- Tests added in test_torch.py for all available devices and dtypes (except torch.bfloat16)

Fixes #24088

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -3874,10 +3874,21 @@ def test_diagflat(self):
...
+
+            for n, m in product([3, 5, 7], repeat=2):                
+                # Construct identity using diagonal and fill
+                res1 = torch.eye(n, m, device=device, dtype=dtype)
+                naive_eye = torch.zeros(n, m, dtype=dtype, device=device)
+                naive_eye.diagonal(dim1=-2, dim2=-1).fill_(1)
+                self.assertEqual(naive_eye, res1)
+
+                # Check eye_out outputs
+                res2 = torch.Tensor().to(device=device, dtype=dtype)
```
Comment (xuhdev) [inline test/test_torch.py]: Would you like to use `res2 = torch.empty(0, device=device, dtype=dtype)`?

Comment (vishwakftw) [reply test/test_torch.py]: Thank you, I've made the change.

File: test/test_torch.py [hunk 2]
```diff
@@ -3874,10 +3874,21 @@ def test_diagflat(self):
         self._test_diagflat(self, dtype=torch.float32, device='cpu')
 
     def test_eye(self):
-        res1 = torch.eye(100, 100)
-        res2 = torch.Tensor()
-        torch.eye(100, 100, out=res2)
-        self.assertEqual(res1, res2)
+        for dtype, device in product(torch.testing.get_all_dtypes(), torch.testing.get_all_device_types()):
```
Comment (xuhdev) [inline test/test_torch.py]: This doesn't seem having `half` tested

Comment (vishwakftw) [reply test/test_torch.py]: get_all_dtypes() contains returns a list that contains torch.float16.

https://github.com/pytorch/pytorch/blob/465b4de9d4378ac58013b56ea45fe30f9beadb39/torch/testing/__init__.py#L89-L91

Comment (xuhdev) [reply test/test_torch.py]: Yes, you are right.

**Review by xuhdev — APPROVED** [review]

File: test/test_torch.py [hunk 3]
```diff
@@ -3874,10 +3874,21 @@ def test_diagflat(self):
...
+            if dtype == torch.bfloat16:
+                continue
+
+            for n, m in product([3, 5, 7], repeat=2):
```
Comment (pietern) [inline test/test_torch.py]: Trailing whitespace

Comment (vishwakftw) [reply test/test_torch.py]: Thank you for pointing this out. I'll push a commit to rectify this.

Comment (pietern) [reply test/test_torch.py]: Thanks, @vishwakftw. I see we ignore it in the `.flake8` configuration.

Let me check if we can enable it easily...

Comment (vishwakftw) [reply test/test_torch.py]: https://github.com/pytorch/pytorch/pull/19828: this was the PR that removed W291 from .flake8.

Comment (pietern) [reply test/test_torch.py]: Thanks! I added #24238 to re-enable it.

**Comment by vishwakftw** [comment]
@pietern is this good to go?

**Comment by gchanan** [comment]
were these also enabled on CUDA?  If not, why?

**Comment by vishwakftw** [comment]
> were these also enabled on CUDA? If not, why?

Yes they were. We also test for them.

**Pull request closed without merging**